Implement bluff support

flackr opened the pull request:
This implements bluffs as a special type of hidden finesse. A bluff connection is originally recognized as a layered finesse with a bluff attribute being marked on the first connection if the conditions for a bluff are met. Once the full connection is known, the validity of the bluff and bluff target can be checked to determine if this should be interpreted as a bluff. This is further adjusted to prefer non-bluff interpretations when those exist and would be ambiguous with a possible bluff.

flackr commented:
Hey, this is still really early, but I am really excited to get the bot bluffing and understanding bluffs so I had a try at implementing it. It seems to be working pretty well so far but I haven't extensively tested it yet.

WillFlame14 commented:
Hey, thanks for taking the initiative! Bluffs are pretty far from where I am now but I think they're definitely vital to playing efficiently, so it'd be really cool if we could get this implemented sooner. :slightly_smiling_face: I know my code is a bit unusual and comment-sparse, but if you have any questions I'd be happy to explain. (I also have [this documentation](https://docs.google.com/document/d/1JMXtNnv3Bw_4Lf6uW_KIllp-Eb7d2wqa_vFaPJzWbDw/edit?usp=sharing), but I haven't updated it in a while.)

WillFlame14 commented:
Any thoughts on the specific implementation of bluffs by always assuming a layered finesse and then cutting off connections after the first? My initial thought was to add a new 'bluff'-type connection instead of a hidden connection (e.g. around line 161 of connecting-cards.js). However, find_connecting() and related would need to take a bluff parameter to ensure that they still find layered finesses when the focused card isn't a valid bluff target, since those functions only search for a single identity without knowledge of the wider context. It does seem like your approach is less intrusive, even if it does some unnecessary work.

flackr commented:
Thanks for the documentation, I'll definitely have to read this over!

> Any thoughts on the specific implementation of bluffs by always assuming a layered finesse and then cutting off connections after the first? My initial thought was to add a new 'bluff'-type connection instead of a hidden connection (e.g. around line 161 of connecting-cards.js).

This is basically what I've done. I've added a bluff attribute to the connection returned by `find_unknown_connecting` so we know whether to cut off the connection. I originally tried just allowing the hidden connection but it requires a lot of changes as the code really isn't expecting the connection to be hidden. It seems to be cleaner overall to have a non-hidden bluff type of connection.

> However, find_connecting() and related would need to take a bluff parameter to ensure that they still find layered finesses when the focused card isn't a valid bluff target, since those functions only search for a single identity without knowledge of the wider context. It does seem like your approach is less intrusive, even if it does some unnecessary work.

Yeah, I have a TODO in the code for this. Right now it will always assume a bluff if the first play in the connection comes from the bluff seat. By only allowing a bluff for the first connection it should ensures the first part of a valid bluff target, that the eventual target most only have one missing link, but should allow non-bluff connections through other cards.

The part that's not yet covered is that it needs to be clear from the play of the bluff card that the next card does not follow it. I haven't quite figured out what this check looks like. I think I might need to compute the chain as if it was not bluffed and then only cut it short if the next card comes from the same hand or is known to not be playable after the bluff play.

The inferred value on the bluffed card is any non-duplicated immediately playable card. I think I also need to add the bluff to waiting_connections so that if its not played the inferred value is updated.

flackr commented:
> Bluffs are pretty far from where I am now but I think they're definitely vital to playing efficiently, so it'd be really cool if we could get this implemented sooner. 🙂

Yeah, I have many people I play with who would happily ignore most of the levels but need their bluffs :grin:. I was even thinking it might be nice to be able to add in bluff support at lower levels without having to know all of the rules up to 11.

RandoPerso commented:
I noticed that the bots sometimes try to bluff through finesse position cards, causing bombing.
(https://hanab.live/shared-replay/1158127#10)
![image](https://github.com/WillFlame14/hanabi-bot/assets/86388449/0eedafc2-b671-4fb7-bb9c-f9f596ecdcd4)
I believe this happens in the rare circumstance that a clue bluffing through finesse position cards would also touch valuable cards. The following two test setups should have the same issue:
```js
const game = setup(HGroup, [
	['xx', 'xx', 'xx', 'xx', 'xx'],
	['p1', 'y4', 'r4', 'g4', 'b4'],
	['r2', 'r3', 'b3', 'g1', 'g1'],
], {
	level: 11,
	play_stacks: [0, 0, 1, 1, 0],
	clue_tokens: 6
});
```
(Alice clues 3 instead of the expected red clue)
```js
const game = setup(HGroup, [
	['xx', 'xx', 'xx', 'xx', 'xx'],
	['p1', 'y4', 'r4', 'g4', 'p4'],
	['r3', 'r4', 'b4', 'g4', 'g1'],
], {
	level: 11,
	play_stacks: [1, 0, 1, 3, 0],
	clue_tokens: 6
});
```
(Alice clues 4 instead of the expected red clue)

RandoPerso commented:
Noticed a few more things:
(https://hanab.live/shared-replay/1158131#2)
After b1 plays, the unknown 2 should become b2 from Cathy's Connecting Principle.
![image](https://github.com/WillFlame14/hanabi-bot/assets/86388449/4f3a374e-4a40-43f8-8271-6a42bbcbde36)

(https://hanab.live/shared-replay/1158131#13)
A queued bluff was attempted (b3 and b4 are already finessed)
![image](https://github.com/WillFlame14/hanabi-bot/assets/86388449/d52ca9d2-58ba-4f85-9e1b-5a6cc658b881)

flackr commented:
@RandoPerso Thanks for pointing out all these additional cases. I've landed fixes for everything called out here and also removed my hack of hard coding the level to 11 and instead updating the max level constant.

All tests are passing :-).

WillFlame14 commented:
It looks like there are some problems with faking a connection as a bluff. See this test:
```js
it(`understands a complex play if the bluff isn't played into`, () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx'],
		['b3', 'g3', 'r3', 'g5'],
		['p1', 'r4', 'b5', 'b2'],
		['r2', 'b3', 'g1', 'y5']
	], {
		level: 11,
		starting: PLAYER.DONALD
	});
	takeTurn(game, 'Donald clues 2 to Cathy');    // 2 save
	takeTurn(game, 'Alice clues 5 to Bob');       // 5 save
	takeTurn(game, 'Bob clues blue to Donald');   // finesse for b1 on us
	takeTurn(game, 'Cathy clues 5 to Donald');    // 5 save

	assert.equal(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order].finessed, true);
	ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order], ['b1']);
});
```
Alice doesn't add a waiting connection for Cathy's p1 after Bob's clue, because a focus possibility for b3 is added with no connections after the code rejects this as a valid bluff (since p1 can't be played into b2). This does seem like a valid bluff though, see [bluffs through already-clued cards](https://hanabi.github.io/level-11#bluffs-through-already-clued-cards).

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -159,15 +160,75 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
 		else if (game.level >= LEVEL.INTERMEDIATE_FINESSES && state.play_stacks[finesse.suitIndex] + 1 === finesse.rank) {
-			// Could be duplicated in giver's hand - disallow hidden finesse
-			if (state.hands[giver].some(c => c.clued && game.players[giver].thoughts[c.order].inferred.has(identity))) {
+			let bluff = false;
```
This can be simplified into one statement without needing to use `let`:
```js
const bluff = game.level >= LEVEL.BLUFFS &&
    connected.length == 1 &&
    ((giver + 1) % state.numPlayers) == reacting
    && !common.thinksLoaded(state, reacting);
```
This pattern also appears on line 347 of own-finesses.js.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -159,15 +160,75 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
+				bluff =
+					connected.length == 1 &&
+					((giver + 1) % state.numPlayers) == reacting &&
+					!common.thinksLoaded(state, reacting);
```
Simply checking if the reacting player is loaded seems too broad. I think the reacting player only cannot have finesses that have room to be layered on them, but I'm not too sure. (I think if they have a card that could be prompted instead they wouldn't even get to this point.)

flackr replied:
Agreed, I think `unknown_plays` is the right variable? I've added this and a test that it can bluff with known plays.

WillFlame14 replied:
I think this is closer, but not quite there yet. `unknown_plays` contains the cards that the player knows are delayed playable without knowing their identity (i.e. the card has more than 1 inference). However, a player can still be bluffed even with unknown plays-- I think the only issue is that the reacting player won't bluff if a finesse on them could be layered. See these tests which currently fail:
```js
it('understands a bluff on top of unknown plays that cannot match', () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx'],
		['r1', 'p3', 'y1', 'y1'],
		['g5', 'b2', 'r3', 'y5'],
		['b4', 'p2', 'g3', 'r5'],
	], {
		level: 11,
		play_stacks: [0, 0, 1, 0, 2],
		starting: PLAYER.DONALD
	});
	takeTurn(game, 'Donald clues 1 to Bob');
	// Since Bob is only queued on 1s, Alice should be able to bluff Bob's p3 using g3.
	takeTurn(game, 'Alice clues 3 to Donald');

	assert.equal(game.common.thoughts[game.state.hands[PLAYER.BOB][1].order].finessed, true);
});

it(`doesn't bluff on top of unknown queued cards`, () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx', 'xx'],
		['r2', 'g1', 'y1', 'y1', 'r4'],
		['p2', 'b2', 'r3', 'y5', 'y4'],
		['g1', 'g2', 'g3', 'g5', 'p4'],
	], {
		level: 11,
		play_stacks: [1, 0, 0, 0, 0],
		starting: PLAYER.DONALD
	});
	takeTurn(game, 'Donald clues red to Cathy');
	// With r2 queued, we cannot bluff the g1.
	const { play_clues } = find_clues(game);
	const bluff_clues = play_clues[2].filter(clue => {
		return clue.type == CLUE.RANK && clue.target == 2 && clue.value == 2 ||
			clue.type == CLUE.COLOUR && clue.target == 2 && (clue.value == 3 || clue.value == 4);
	});
	assert.equal(bluff_clues.length, 0);
});
```

Maybe something like `!state.hands[reacting].some((c, index) => common.thoughts[c.order].finessed && state.hands[reacting].some((c2, index2) => index2 > index && !c2.clued));`? Which passes both tests above, but doesn't take negative information on cards into consideration. It's kind of a weird situation, for example, if you're playing into a y2 layered finesse (i.e. not from slot 1), and the only other unclued card is a chop moved card that is either 4 or 5, and you are called to play g1. Then you know that you can just play slot 1, but this expression doesn't account for that.

But it could be even more awkward if the "y2" card could be g1 and the other unclued card could be y2, because then the layer is a possibility again, and the bluff can't be given. I think it's something like if the union of both cards have both identities that are being looked for, then a bluff can't be given, but I'm worried that the finessed card may not even have a single identity (e.g. if they are in a superposition of finesses due to an unknown self-component, but maybe this situation isn't possible). It's difficult to think about, not sure if you have any ideas or if you think we should just fail to interpret these situations. :sweat_smile:

flackr replied:
I added these tests and they pass with my latest patch. The condition now is that they think their unknown card matches the requested identity.

WillFlame14 replied:
Hmm, that still doesn't sound quite right to me. Players don't "know" their own cards other than from a rewind, so this would pretty much always be false. In fact, removing this expression completely still passes all the current tests.

Currently, it seems that any cards that might be played instead will be added to `connected` first, making the length greater than 1 and failing the earlier condition anyways. I think it makes sense since playable/prompt connections are always prioritized over finesse connections, and if the prompt isn't playable then it will trigger the wrong prompt line. But I'm not sure, was this line necessary before?

WillFlame14 commented on test/h-group/level-11.js:
```diff
@@ -0,0 +1,299 @@
+		// Bob's slot 1 could be any of the playable 3's.
+		assert.equal(game.common.thoughts[game.state.hands[PLAYER.BOB][0].order].finessed, false);
+		// Cathy's slot 2 will be known to be r3 or r4 until Bob reacts.
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.CATHY][1].order], ['r3']);
```
The comment here says that Cathy's slot 2 is [r3,r4], but the code only checks for r3. I'm not sure which was the intention, but it seems like it should be [r3,r4].

flackr replied:
This seems to be consistent with how clues that look like finesses are tracked, every player who knows notes the eventual result. I agree it feels like it should be [r3, r4], but it should be the same as regular finesses, so if we want to change this I'd expect to do it for both. I suspect it's better for everyone to track what will be known so that they don't worry about the other possibility for future clues.

WillFlame14 commented on test/h-group/level-11.js:
```diff
@@ -0,0 +1,299 @@
+		// Alice's slot 1 should is not assumed
+		assert.equal(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order].finessed, false);
+		// Bob will know this isb4.
+		ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.BOB][0].order], ['b3', 'b4']);
```
The comment here says that Bob should know b4, but I think the code is correct in only assuming [b3,b4] on Bob's slot 1.

flackr replied:
Updated the comments, by "will know" I meant after the play happens. Of course if the play doesn't happen Bob will know that it's b3.

WillFlame14 commented:
Also, some things that may not have been obvious with my test suite:
- By default, the play stacks are all empty and Alice is the starting player. If either of these are true, then you don't need to specify them in `setup()`'s options.
- Since only Alice's hand is unknown, `takeTurn()` only needs to specify slot information when dealing with Alice's hand (or when the card could be ambiguous). Strings like `'Bob clues green to Donald'` and `'Cathy plays g1'` don't need slot information, since that can be annoying to maintain if the test changes. Similarly, the identity of the drawn card when Alice plays/discards doesn't need to be specified.

flackr commented:
> It looks like there are some problems with faking a connection as a bluff. See this test:
> 
> ```js
> it(`understands a complex play if the bluff isn't played into`, () => {
> 	const game = setup(HGroup, [
> 		['xx', 'xx', 'xx', 'xx'],
> 		['b3', 'g3', 'r3', 'g5'],
> 		['p1', 'r4', 'b5', 'b2'],
> 		['r2', 'b3', 'g1', 'y5']
> 	], {
> 		level: 11,
> 		starting: PLAYER.DONALD
> 	});
> 	takeTurn(game, 'Donald clues 2 to Cathy');    // 2 save
> 	takeTurn(game, 'Alice clues 5 to Bob');       // 5 save
> 	takeTurn(game, 'Bob clues blue to Donald');   // finesse for b1 on us
> 	takeTurn(game, 'Cathy clues 5 to Donald');    // 5 save
> 
> 	assert.equal(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order].finessed, true);
> 	ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][0].order], ['b1']);
> });
> ```
> 
> Alice doesn't add a waiting connection for Cathy's p1 after Bob's clue, because a focus possibility for b3 is added with no connections after the code rejects this as a valid bluff (since p1 looks like it can be played into b2). This does seem like a valid bluff though, see [bluffs through already-clued cards](https://hanabi.github.io/level-11#bluffs-through-already-clued-cards).

Agree this is a valid bluff, and you uncovered a bit of laziness on my part about identifying when a bluff is valid where I only checked that the immediate next card is known incompatible. I fixed that and it does add the connection, but I think your test case is not correct because the 5 save is critical isn't it? I.e. it doesn't seem to consider that a problem for delaying the finesse play. Maybe this is a problem since usually bluffs take priority to prevent mis-plays though in this case there's no danger of that.

I changed the y5 to a y1 and had Cathy discard and it now passes, but bluffs probably should take priority over everything else, and also not be given if a critical save is needed.

flackr reviewed (commented):
Thanks for the review, I think I've cleaned up the testing style as you hinted at and addressed all of the cases mentioned.

WillFlame14 commented on test/h-group/level-11.js:
```diff
@@ -0,0 +1,345 @@
+			starting: PLAYER.DONALD
+		});
+		takeTurn(game, 'Donald clues red to Bob');
+		// r2 is known to be queued, and would be played over finesse slot.
```
It doesn't look like this test matches the name, assuming I understand your intent. Only r1 is queued in Bob's hand, and the reason why 2 to Cathy is not permitted is because g1 would look like it connects to b2, not because r1 would be played instead.

Maybe these are closer to what you were going for? The code passes both of them.
```js
it(`doesn't bluff on top of colour-clued cards which might match bluff`, () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx', 'xx'],
		['r1', 'r4', 'y1', 'y1', 'y5'],
		['p4', 'b2', 'r3', 'y5', 'y4'],
	], { level: 11 });

	takeTurn(game, 'Alice clues red to Bob');
	takeTurn(game, 'Bob plays r1', 'g1');
	takeTurn(game, 'Cathy clues 5 to Bob');

	// Alice cannot use r3 to bluff Bob's g1, as r4 would play instead.
	const { play_clues } = find_clues(game);
	assert.ok(!play_clues[PLAYER.CATHY].some(clue =>
		(clue.type === CLUE.RANK && clue.value === 3) ||
		(clue.type === CLUE.COLOUR && clue.value === COLOUR.RED)));
});

it(`doesn't bluff on top of rank-clued cards which might match bluff`, () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx', 'xx'],
		['y2', 'r2', 'y1', 'y1', 'y5'],
		['g4', 'b2', 'p3', 'y5', 'y4'],
	], {
		level: 11,
		play_stacks: [0, 1, 0, 0, 1]
	});

	takeTurn(game, 'Alice clues 2 to Bob');
	takeTurn(game, 'Bob plays y2', 'g1');
	takeTurn(game, 'Cathy clues 5 to Bob');

	// Alice cannot use p3 to bluff Bob's g1, as r2 would play instead.
	const { play_clues } = find_clues(game);
	assert.ok(!play_clues[PLAYER.CATHY].some(clue =>
		(clue.type === CLUE.RANK && clue.value === 3) ||
		(clue.type === CLUE.COLOUR && clue.value === COLOUR.PURPLE)));
});
```

flackr replied:
Thanks, yeah this is exactly what I was thinking, added!

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -159,15 +160,90 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
+	// Determine the next play if this is a bluff.
+	let next_play = connections.findIndex(connection => connection.card == promised[0]) + 1;
+	if (next_play == 0)
+		next_play++;
```
Is this part necessary? It seems like if there are no connections that are promised, that's because the only connection found was a bluff, which means that `bluff_possible` can't be false and we can skip doing the next loop. (It seems like incrementing `next_play` actually skips the loop too, but it isn't obvious that this is the case.) Then `next_play` could be `const`, and we could simplify this into:
```js
const next_play = connections.findIndex(connection => connection.card == promised[0]) + 1;
let bluff_possible = next_play === 0 ||
    connections.every((conn, index) => index < next_play || (!conn.hidden && conn.type !== 'finesse'));
```
I'm not confident though, correct me if I'm wrong.

flackr replied:
In fact, it's just plain not possible to not find the first promised card. It is either the bluff card or some other card in connections. Fixed!

flackr commented:
I uploaded the work in progress in case anyone wants to follow along, but there is a fundamental issue right now. We should only assume a bluff once we've eliminated the possibility for any non-bluff connections. So what this means is that the current `resolve_bluff` function should only be determining if a bluff is possible for a given connection, but once we have all the connections then if we find a non bluff connection we can convert all of the remaining potential bluff connections to non-bluff connections.

WillFlame14 commented:
Do you have an example that explains the problem? I'm having a bit of trouble following. Since a bluff must resolve immediately, it seems fine if we always assume a bluff is possibly occurring (unless it's on us and we can see a valid finesse), and rewind after if it doesn't. Not playing into a finesse/bluff causes that card to be ignored when computing connections in a rewind, so it should find real finesses after that. Then we don't need to worry about computing everything the first time around.

I've made some changes in master to how ignoring cards works and fixed a small bug where play stacks didn't reset to their original values in a test after rewinding, so if it just seemed like a strange issue then maybe merging in master will help.

flackr commented:
I think it might also be a good idea to run tests at the max level unless they specifically require not including later conventions as it would increase coverage of things like whether the bluff support breaks any tests in earlier levels.

WillFlame14 commented:
Okay, I think I see now. If you haven't already thought of a solution, I think you could make `resolve_bluff()` simply return a boolean indicating whether a bluff is possible, and add focus possibilities depending on its result. If it's possible, add both the full connection and a bluff connection containing just the first card. If it's not, add nothing if the connection has no promised card, or the full connection otherwise. Then if the bluff connections are filtered out later on, the multiple finesse connections still exist. If both still exist, they can be filtered using [Occam's Razor](https://hanabi.github.io/level-11#clue-interpretation--occams-razor).

> I think it might also be a good idea to run tests at the max level unless they specifically require not including later conventions as it would increase coverage of things like whether the bluff support breaks any tests in earlier levels.

This is a good idea, I guess I'd make the default level be max and leave specifications for tests that need to be done at a specific level.

WillFlame14 commented:
Also, changing logConnection to indicate bluffs would be nice to help understand logs:
```js
export function logConnection(connection) {
	const { type, reacting, identities, card } = connection;
	const identity = identities.length === 1 ? logCard(identities[0]) : `[${identities.map(logCard)}]`;
	const logType = type === 'finesse' ? (connection.bluff ? 'bluff' : 'finesse') : type;

	return `${card.order} ${identity} ${logType} (${globals.game.state.playerNames[reacting]})${connection.certain ? ' (certain)' : ''}`;
}```

flackr commented:
> Okay, I think I see now. If you haven't already thought of a solution, I think you could make `resolve_bluff()` simply return a boolean indicating whether a bluff is possible, and add focus possibilities depending on its result. If it's possible, add both the full connection and a bluff connection containing just the first card. If it's not, add nothing if the connection has no promised card, or the full connection otherwise. Then if the bluff connections are filtered out later on, the multiple finesse connections still exist. If both still exist, they can be filtered using [Occam's Razor](https://hanabi.github.io/level-11#clue-interpretation--occams-razor).

Thanks for the suggestion, but the idea I had worked! I was able to get the new tests and all the other ones working as expected :-). I could revisit trying to add both connection possibilities but it is repeating a lot of calculations. It may be that this is just better in the long run though. I think there's a few other comments you've made I've yet to address but this is looking good now.

flackr commented:
> I think there's a few other comments you've made I've yet to address but this is looking good now.

I think I've addressed all issues and suggested changes now, apologies if I missed something.

WillFlame14 commented:
Thanks! I think there are still a couple problems in connections, though:
```js
it(`computes connections correctly`, () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx'],
		['y2', 'r1', 'r2', 'y4'],
		['y3', 'p2', 'y1', 'r4'],
		['g5', 'y1', 'p4', 'b5']
	], { level: 11 });

	takeTurn(game, 'Alice clues red to Bob');
	takeTurn(game, 'Bob plays r1', 'b1');
	takeTurn(game, 'Cathy clues yellow to Donald');
	takeTurn(game, 'Donald clues green to Alice (slots 3,4)');

	takeTurn(game, 'Alice clues 3 to Cathy');

	// Simplest interpretations: r2 (Bob) prompt, b1 (Bob) -> y2 (Bob) layered finesse, b1 (Bob) -> b2 (Cathy) self-finesse
	ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.CATHY][0].order], ['r3', 'y3', 'b3']);
});
```

The relevant log output:
```
Turn 5: Alice clues 3 to Cathy 
severity 1 
found connections: [5 r2 prompt (Bob) -> r3?] 
found connections: [14 y1 known (Donald) -> 16 b1 finesse (Bob) -> 7 y2 finesse (Bob) -> y3?] 
disallowed hidden finesse on b1 16, true g2 could be duplicated in giver's hand 
found connections: [0 g1 known (Alice) -> g2?] 
found connections: [16 b1 finesse (Bob) -> b2?] 
disallowed finesse on y1, playable already clued elsewhere 
found connections: [16 b1 finesse (Bob) -> 10 p2 finesse (Cathy) -> p3?] 
focus possible: [ 'r3', 'y3', 'p3' ] 
symmetric connections [
  '[5 r2 prompt (Bob) -> r3?]',
  '[14 y1 known (Donald) -> 16 b1 bluff (Bob) -> 7 y2 finesse (Bob) -> y3?]',
  '[0 g1 known (Alice) -> 16 b1 bluff (Bob) -> g3?]',
  '[16 b1 finesse (Bob) -> 10 b2 finesse (Cathy) -> b3?]'
] 
final inference on focused card r3,y3,g3,b3,p3
```

Note that the purple connection is impossible since p1 doesn't exist, and two of the symmetric connections contain a "bluff" in the middle (and it also finds the impossible purple one). At least for the purple connection, it looks like b1 originally used to be a bluff, and then it was "converted" into a finesse.

I'm not entirely sure what you've done with making bluffs "hidden" or not, but it seems like that's at least part of the problem: `find_connecting()` sets a bluff connection to not be hidden (line 325 of connecting-cards.js), but then `resolve_bluff()` expects the bluff connection to be hidden in order to reject the connection (line 231 of connecting-cards.js). I'm not sure why the symmetric connections are so strange, though.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/interpret-clue.js:
```diff
@@ -297,6 +302,32 @@ export function interpret_clue(game, action) {
+				// Remove everything after the bluff play to the non-hidden play as they won't
+				// play after the bluff play.
+				const next_visible_connection = conn.connections.findIndex(c => !c.bluff && !c.hidden);
+				conn.connections.splice(1, next_visible_connection);
```
Should this be `next_visible_connection - 1`? If the next real connection is at index 2, then I think we only want to remove 1 element, but this will remove 2.

flackr replied:
> Should this be `next_visible_connection - 1`? If the next real connection is at index 2, then I think we only want to remove 1 element, but this will remove 2.

No, I think we do want to remove the "real" connection if the first card is a legitimate bluff, as it can't be known that this real connection exists. E.g. if you are bluffed to play an r2 and it doesn't match the real one won't be known.

> Will there be later connections we need to preserve in the first place (maybe when bluffing through already-clued cards)?

Yes, my thinking was that a bluff can reveal to other players that they have cards that resulted in the bluff being valid. E.g. learning that your red card is in fact a red 3 to make a bluff for the r2 work by clueing an r4.

flackr replied:
I think the only real case where this could happen is when it comes from the same hand. E.g.

Alice clues red to Cathy,
Bob's hand is b1, r1

Bob is going to stop playing after the b1, so we want to erase the r1 connection.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/interpret-clue.js:
```diff
@@ -297,6 +302,32 @@ export function interpret_clue(game, action) {
 		}
 
+		// If there's a valid connection that doesn't require a bluff, a bluff is not a valid interpretation.
+		const no_bluff_connections = all_connections.filter(connection => connection.connections.length > 0 && !connection.connections[0].bluff);
```
If we don't need `no_bluff_connections`, we can just use `.some()` instead of `.filter()` so that it'll exit early once it finds one.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -159,17 +160,82 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
+			game.players[state.ourPlayerIndex].thoughts[bluffCard.order].inferred.filter(c => state.isPlayable(c)) :
+			[bluffCard];
+		for (let i = 1; i < promised.length; ++i) {
+			const known_bluff_cards = bluff_identies.filter(identity => {
```
Since we're not using `known_bluff_cards`, we can use `.some()` instead of `.filter()` here. Also, I think there's a spelling mistake in "bluff_**identities**".

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -159,17 +160,82 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
+	if (!bluff_possible) {
+		// If a bluff is not possible, we only have a valid connection if the real card was found.
+		// If it is in our hand, we have to assume it is the promised card.
+		if (connections[0].card == promised[0] && connections[0].hidden)
```
If you're changing to compare cards using order, then there's another one here. :slightly_smiling_face:

Throughout your PR, either way should be fine since all of this is done at the same time. What can occur in other places is that the game gets cloned during a rewind/hypothetical, and then two objects don't have the same reference anymore. Using the order doesn't have this problem, so I've just gotten used to doing it that way.

flackr commented:
> The relevant log output:

What's the trick to get the log output from test runs?

> Note that the purple connection is impossible since p1 doesn't exist, and two of the symmetric connections contain a "bluff" in the middle. At least for the purple connection, it looks like b1 originally used to be a bluff, and then it was "converted" into a finesse.

> I'm not entirely sure what you've done with making bluffs "hidden" or not, but it seems like that's at least part of the problem: `find_connecting()` sets a bluff connection to not be hidden (line 325 of connecting-cards.js), but then `resolve_bluff()` expects the bluff connection to be hidden in order to reject the connection (line 231 of connecting-cards.js). I'm not sure why the symmetric connections are so strange, though.

Let me explain what I'm trying to do, in case you have better suggestions for how to manage this.

Bluffs are "hidden" when they are the first card in a layered finesse. This is so that if we later determine to not treat the connection as a bluff we can convert it back into a hidden layered finesse:

```
			// Convert possible bluff connections to non-bluff connections.
			all_connections = all_connections.map(conn => {
				if (!conn.connections[0]?.bluff)
					return conn;
				// If not a hidden connection, and we know the bluff card doesn't match, the real card wasn't found.
				if (!conn.connections[0].hidden && !conn.connections[0].identities.find(i => conn.connections[0].card.matches(i, {assume: true})))
					return null;
				conn.connections[0].bluff = false;
				return conn;
			}).filter(conn => !!conn);
```

If we don't find the real card, then we update the bluff card to be non-hidden so that we know in the sequence of connections that the real card wasn't found and we should treat the bluff card as being the "real card" we were after - even though we know it isn't the correct one:

```
		// If we don't find the actual card, a bluff of the first card is still a valid interpretation.
		if (connections.length > 0 && connections.at(0).bluff && connections.at(-1).hidden) {
			// Remove all of the hidden plays after the bluff and treat the bluff as the known target.
			connections.splice(1, connections.length - 1);
			connections[0].hidden = false;
		}
```

Still looking into the test case to see exactly what went wrong.

WillFlame14 commented:
> What's the trick to get the log output from test runs?

There's a line near the top of every test file where the logging level is set to ERROR, to prevent the console from being spammed. I comment out that line as well as all the other tests in that file, and then run that file on its own using `node <path-to-file>`. There's also a way to make node's test runner [only run a particular test](https://nodejs.org/api/test.html#only-tests), but I find commenting things out easier to work with.

flackr commented:
So I fixed the test case you provided, and also discovered and fixed many other issues in play testing :grin:

I also went through some of the cases in https://hanabi.github.io/level-11/ and made sure it understood them (spoiler it didn't but does now).

I *think* this might be ready?

flackr commented:
Other than that it definitely isn't giving the bluff clues enough value and prefers to direct clue a single play instead :(, I'll see if i can figure out why.

flackr commented:
> Other than that it definitely isn't giving the bluff clues enough value and prefers to direct clue a single play instead :(, I'll see if i can figure out why.

The problem is the following from clue_finder.js:
```
// Clues where the focus isn't playable or that cause chop moves aren't plays/stalls
if ((playables.length > 0 && !playables.some(({ card }) => card.order === focused_card.order)) ||
		(playables.length === 0 && chop_moved.length > 0) ||
		isTrash(state, me, focused_card, focused_card.order)) {
	logger.highlight('yellow', 'invalid play clue');
	continue;
}
```

Bluffs are exactly clues that cause the focus not to be playable, but should still be considered. I think we just need to check that if the focus is not playable there is a waiting bluff connection that will reveal it.

flackr commented:
I think I failed to merge some of the changes on master correctly before (long lived branches are hard), the test failures should be fixed now.

WillFlame14 commented:
I hope we're almost there! :sweat_smile: I noticed some weird behaviour still:

```js
it(`understands a clandestine finesse`, () => {
	const game = setup(HGroup, [
		['xx', 'xx', 'xx', 'xx'],
		['y4', 'y5', 'y1', 'y1'],
		['g3', 'g3', 'b2', 'b1'],
		['g1', 'r1', 'r3', 'y2']
	], {
		level: 11,
		starting: PLAYER.CATHY
	});

	takeTurn(game, 'Cathy clues 3 to Alice (slot 4)');

	// Alice's slot 4 should be r3 as a Clandestine Finesse (no 3 is a valid bluff target, and both g3s are in Cathy's hand).
	ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][3].order], ['r3']);
});
```

Relevant log output:
```
Turn 1: Cathy clues 3 to Alice 
severity 1 
found connections: [15 g1 bluff (Donald) -> r2?] 
found connections: [15 g1 bluff (Donald) -> y2?] 
found connections: [15 g1 bluff (Donald) -> g2?] 
found connections: [15 g1 bluff (Donald) -> b2?] 
found connections: [15 g1 bluff (Donald) -> p2?] 
focus possible: [] 
card (unknown) order 0 doesn't match any inferences! currently r3,y3,g3,b3,p3 
found connections: [15 g1 bluff (Donald) -> 3 r2 finesse (Alice) -> r3?] 
found connections: [15 g1 bluff (Donald) -> 3 y2 finesse (Alice) -> y3?] 
found connections: [15 g1 bluff (Donald) -> 3 g2 finesse (Alice) -> g3?] 
found connections: [15 g1 bluff (Donald) -> 3 b2 finesse (Alice) -> b3?] 
found connections: [15 g1 bluff (Donald) -> 3 p2 finesse (Alice) -> p3?] 
selecting inferences [ 'r2', 'y3', 'g3', 'b3', 'p3' ]
```

For some reason, it looks like it's still trying to connect finesses on bluffs.

I'll find some time before we merge this to go through the other tests and make sure they still pass at this level.

WillFlame14 commented:
Also, I should reveal my secret on finding these: the bot has a self-play feature that I forgot to document on the README.
For example, `npm run self-play -- players=4 games=20 level=11` will create a `seeds/` folder with JSON replays of each seed, and I typically look in games that scored < 20/25 or struck out. This is a lot faster than going through hanab.live's UI to play each game, and you can still use hanab.live to load the selected replays afterward. You can also provide an integer seed (e.g. `seed=1939`) and it will start at that seed.

The `replay` feature also allows reading from a file, e.g. `npm run replay -- file=seeds/0.json level=11 index=0` will simulate as Alice at level 11.

flackr commented:
> I hope we're almost there! 😅 I noticed some weird behaviour still:
> 
> ```js
> it(`understands a clandestine finesse`, () => {
> 	const game = setup(HGroup, [
> 		['xx', 'xx', 'xx', 'xx'],
> 		['y4', 'y5', 'y1', 'y1'],
> 		['g3', 'g3', 'b2', 'b1'],
> 		['g1', 'r1', 'r3', 'y2']
> 	], {
> 		level: 11,
> 		starting: PLAYER.CATHY
> 	});
> 
> 	takeTurn(game, 'Cathy clues 3 to Alice (slot 4)');
> 
> 	// Alice's slot 4 should be r3 as a Clandestine Finesse (no 3 is a valid bluff target, and both g3s are in Cathy's hand).
> 	ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][3].order], ['r3']);
> });
> ```

Ruling out the g3 should be done upstream (outside of bluff determination), so I filed https://github.com/WillFlame14/hanabi-bot/issues/209 for this.

I did fix the failure to recognize the clandestine finesse, requiring all sorts of tweaks - though arguably improvements - to other behaviors to ensure all the other test cases pass :).

flackr commented:
I wonder how much logging would be worth adding to help with this, it's a tough balance between useful logging and making it too verbose, but it might be worth logging a single line for why resolve_bluff determines a potential bluff is either possible or not.

flackr commented:
> Ruling out the g3 should be done upstream (outside of bluff determination), so I filed https://github.com/WillFlame14/hanabi-bot/issues/209 for this (I verified it was broken without this PR).

I started to work on this and quickly realized the test expectation is wrong. It's not common knowledge that Cathy has both g3's so it shouldn't exclude g3 in `game.common.thoughts`, but is correctly excluded in `game.players[PLAYER.ALICE].thoughts`.

WillFlame14 commented:
> I wonder how much logging would be worth adding to help with this, it's a tough balance between useful logging and making it too verbose, but it might be worth logging a single line for why resolve_bluff determines a potential bluff is either possible or not.

I think it'd be nice to have a line reasoning why a bluff is impossible. :+1: Would be helpful for debugging when a bluff should've been interpreted but ended up not being.

This one is really strange. The self-finesse doesn't work if Donald gives it in a 4p game, but does if Cathy gives it in a 3p game. Maybe there's something wrong with bluff seat indexing?

```js

flackr commented:
> > I wonder how much logging would be worth adding to help with this, it's a tough balance between useful logging and making it too verbose, but it might be worth logging a single line for why resolve_bluff determines a potential bluff is either possible or not.
> 
> I think it'd be nice to have a line reasoning why a bluff is impossible. 👍 Would be helpful for debugging when a bluff should've been interpreted but ended up not being.

I've added some logging for some of the interesting edge cases, hopefully it's not too verbose.

> This one is really strange. The self-finesse doesn't work if Donald gives it in a 4p game, but does if Cathy gives it in a 3p game. Maybe there's something wrong with bluff seat indexing?
> 
> ```js
> it('rank connects on a self-finesse', () => {
> 	const game = setup(HGroup, [
> 		['xx', 'xx', 'xx', 'xx'],
> 		['b4', 'p1', 'y1', 'p4'],
> 		['r3', 'p3', 'p1', 'y2'],
> 		['b3', 'y2', 'g4', 'r5']
> 	], {
> 		starting: PLAYER.DONALD,
> 		level: 11
> 	});
> 
> 	takeTurn(game, 'Donald clues 2 to Alice (slot 3)');
> 	takeTurn(game, 'Alice plays g1 (slot 1)');
> 
> 	// Slot 3 should be g2.
> 	ExAsserts.cardHasInferences(game.common.thoughts[game.state.hands[PLAYER.ALICE][2].order], ['g2']);
> });
> ```

So I guess I fixed that on my crusade to make sure that it passes all previous level tests (testing with https://github.com/WillFlame14/hanabi-bot/pull/211 applied)

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -149,40 +151,122 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
+		const bluff = game.level >= LEVEL.BLUFFS &&
+			firstPlay &&
+			((giver + 1) % state.numPlayers) == reacting &&
+			!state.hands[reacting].some(c => game.players[reacting].thoughts[c.order].finessed && game.players[reacting].thoughts[c.order].possible.some(id => id.matches(identity)));
```
Using `possible.has(identity)` is much faster here. The array support in IdentitySet is mainly for situations where transformations are needed.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/connecting-cards.js:
```diff
@@ -149,40 +151,122 @@ function find_unknown_connecting(game, giver, target, reacting, identity, connec
+		// as known by the player who would play next after the bluff play.
+		let known_bluff = false;
+		const next_player = state.hands.findIndex(hand => hand.find(c => c.order == promised[1].order));
+		const bluff_identities = bluffCard.suitIndex == -1 ?
```
This is probably okay currently, but in the future I want to leave open the possibility of adding partial information to cards as they become known, so I would prefer something like `bluffCard.identity() === undefined` here.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/focus-possible.js:
```diff
@@ -72,14 +74,25 @@ function find_colour_focus(game, suitIndex, action) {
 			// Even if a finesse is possible, it might not be a finesse
-			focus_possible.push({ suitIndex, rank: next_rank, save: false, connections: Utils.objClone(connections), interp: CLUE_INTERP.PLAY });
+			promised.push(focused_card);
+			const possible_connections = resolve_bluff(game, giver, target, Utils.objClone(connections), promised, { suitIndex, rank: next_rank });
```
I'd prefer not to clone `connections` unnecessarily, since that can be expensive. It looks like the only modifications to `connections` are near the end of `resolve_bluff()`, where either the first connection has its `bluff` property changed to `false` or a few connections are spliced from the array. For the former I would `slice()` the array, clone the first connection and replace that, and for the latter you can use `connections.toSpliced()` instead.

WillFlame14 commented on src/conventions/h-group/clue-interpretation/focus-possible.js:
```diff
@@ -41,17 +41,19 @@ function find_colour_focus(game, suitIndex, action) {
 		// Note that a colour clue always looks direct
-		const ignoreOrders = game.next_ignore[next_rank - old_play_stacks[suitIndex] - 1]?.filter(i =>
-			i.inference === undefined || i.inference.suitIndex === suitIndex).map(i => i.order);
+		const ignoreOrders = game.next_ignore[next_rank - old_play_stacks[suitIndex] - 1]?.map(i => i.order);
```
Is there a reason why the extra filter was removed? I think this was to prevent incorrectly ignoring cards when the inference of the focused card is different from the case that's different from the one being rewinded, even on the same card and at the same connection index. I might not have written a test case for it though, I'll see if I can recall the conditions and whether they're necessary.

flackr replied:
Yes, this was failing to ignore bluffed cards resulting in it seeing the same bluff connection when the bluff wasn't played into on a rewind and replay. There are tests covering this, I don't remember which one off the top of my head but can check.

flackr replied:
Hmm ... I'm not sure anymore. I know I did this to fix one of the earlier level tests that was failing but perhaps one of the other things I fixed resulted in that test working. I've restored the filter to the way it was before.

flackr replied:
I suspect this change fixed the issue:

https://github.com/WillFlame14/hanabi-bot/pull/184/files#diff-139b83de75bd680f692b861b2b2cda2484860633d9397623d1de63322647aa86R75

As it results in it no longer thinking the bluff in its own hand in "someone else having the connecting card" in the replay. This was necessary for the test `plays queued finesses in the right order` which expects that it knows after the b1 plays that its next card is connecting.

WillFlame14 commented:
It looks like it still performs bluffs very rarely. I think it's because bluffs currently don't count as "finesses", so they're treated with the same efficiency as a normal play clue. Modifying the below in `playables_result()` in clue-result.js seems to fix this and make it perform bluffs much more often.

```js
// Only count unknown playables if they actually go on top of the stacks
if (state.deck[order].rank > player.hypo_stacks[state.deck[order].suitIndex]) {
	if (hypo_player.thoughts[order].finessed)
		finesses.push({ playerIndex, card: hypo_player.thoughts[order] });

	playables.push({ playerIndex, card: hypo_player.thoughts[order] });
}
```

Other than that, this is looking pretty good!

flackr commented:
> It looks like it still performs bluffs very rarely. I think it's because bluffs currently don't count as "finesses", so they're treated with the same efficiency as a normal play clue. Modifying the below in `playables_result()` in clue-result.js seems to fix this and make it perform bluffs much more often.
> 
> ```js
> // Only count unknown playables if they actually go on top of the stacks
> if (state.deck[order].rank > player.hypo_stacks[state.deck[order].suitIndex]) {
> 	if (hypo_player.thoughts[order].finessed)
> 		finesses.push({ playerIndex, card: hypo_player.thoughts[order] });
> 
> 	playables.push({ playerIndex, card: hypo_player.thoughts[order] });
> }
> ```
> 
> Other than that, this is looking pretty good!

Thanks, it's working much better with this change to the scoring!

WillFlame14 reviewed (approved):
I think this is good to merge now, unless you have anything else left to add. :+1: I'll update the README and version after.

flackr commented:
I think it's in a pretty good shape and should be good to land. Thanks for all the review passes, suggestions and added test cases! Also thanks for the cool bot! I really like the way it rewinds to infer what really happened when it makes mistakes.

flackr commented:
Note even with approval I don't have permission to merge the PR, you'll have to do it for me.
